service/history/queues/scheduler_monitor.go

```diff
@@ -0,0 +1,214 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package queues
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/clock"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/tasks"
+	"go.temporal.io/server/common/util"
+)
+
+var (
+	_ tasks.Monitor[Executable] = (*schedulerMonitor)(nil)
+	_ tasks.Monitor[Executable] = (*noopSchedulerMonitor)(nil)
+)
+
+var noopScheduleMonitor = &noopSchedulerMonitor{}
+
+var defaultSchedulerMonitorOptions = schedulerMonitorOptions{
+	aggregationCount:    100,
+	aggregationDuration: time.Second * 30,
+}
+
+type (
+	schedulerMonitor struct {
+		taskChannelKeyFn  TaskChannelKeyFn
+		namespaceRegistry namespace.Registry
+		timeSource        clock.TimeSource
+		metricsHandler    metrics.MetricsHandler
+		options           schedulerMonitorOptions
+
+		status     int32
+		shutdownCh chan struct{}
+
+		sync.Mutex
+		scheduleStats map[TaskChannelKey]*scheduleStats
+	}
+
+	schedulerMonitorOptions struct {
+		aggregationCount    int
+		aggregationDuration time.Duration
+	}
+
+	scheduleStats struct {
+		lastStartTime time.Time
+		totalLatency  time.Duration
+		numStarted    int
+
+		taggedMetricsHandler metrics.MetricsHandler
+		lastEmissionTime     time.Time
+	}
+
+	noopSchedulerMonitor struct{}
+)
+
+func newSchedulerMonitor(
+	taskChannelKeyFn TaskChannelKeyFn,
+	namespaceRegistry namespace.Registry,
+	timeSource clock.TimeSource,
+	metricsHandler metrics.MetricsHandler,
+	options schedulerMonitorOptions,
+) *schedulerMonitor {
+	return &schedulerMonitor{
+		taskChannelKeyFn:  taskChannelKeyFn,
+		namespaceRegistry: namespaceRegistry,
+		timeSource:        timeSource,
+		metricsHandler:    metricsHandler,
+		options:           options,
+
+		status:     common.DaemonStatusInitialized,
+		shutdownCh: make(chan struct{}),
+
+		scheduleStats: make(map[TaskChannelKey]*scheduleStats),
+	}
+}
+
+func (m *schedulerMonitor) Start() {
+	if !atomic.CompareAndSwapInt32(&m.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	go m.metricEmissionLoop()
+}
+
+func (m *schedulerMonitor) Stop() {
+	if !atomic.CompareAndSwapInt32(&m.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	close(m.shutdownCh)
+}
+
+func (m *schedulerMonitor) RecordStart(executable Executable) {
+	startTime := m.timeSource.Now()
```

Comment:
should we have startTime after obtain lock?

Reply:
By the time this method is invoked, the task has already been polled from the task channel and started processing, so I think the start time should be recorded before obtaining the lock (lock latency is not part of queue schedule latency).